File: python/test/t_ConditionedGaussianProcess_std.py
#! /usr/bin/env python

import openturns as ot
import openturns.testing as ott
import openturns.experimental as otexp


def build_kriging_result(inputSample, outputSample, covarianceModel, basis):
    """
    From data & covariance model, build a kriging result
    """
    algo = ot.KrigingAlgorithm(inputSample, outputSample, covarianceModel, basis)
    algo.setOptimizeParameters(False)  # do not optimize hyper-parameters
    algo.run()
    result = algo.getResult()
    return result


def build_gpr_result(inputSample, outputSample, covarianceModel, basis):
    """
    From data & covariance model, build a Gaussian Process Regression result
    """
    fitter_algo = otexp.GaussianProcessFitter(
        inputSample, outputSample, covarianceModel, basis
    )
    fitter_algo.setOptimizeParameters(False)  # do not optimize hyper-parameters
    fitter_algo.run()
    fitter_result = fitter_algo.getResult()
    gpr_algo = otexp.GaussianProcessRegression(fitter_result)
    gpr_algo.run()
    gpr_result = gpr_algo.getResult()
    return gpr_result


ot.TESTPREAMBLE()

ot.PlatformInfo.SetNumericalPrecision(3)
# GP use case
inputDimension = 2

# Learning data
levels = [8.0, 5.0]
box = ot.Box(levels)
inputSample = box.generate()
# Scale each direction
inputSample *= 10


model = ot.SymbolicFunction(["x", "y"], ["cos(0.5*x) + sin(y)"])
outputSample = model(inputSample)

# Validation data
sampleSize = 10
inputValidSample = ot.JointDistribution(2 * [ot.Uniform(0, 10.0)]).getSample(sampleSize)
outputValidSample = model(inputValidSample)

# 2) Definition of exponential model
# The parameters have been calibrated using TNC optimization
# and AbsoluteExponential models
covarianceModel = ot.SquaredExponential([7.63, 2.11], [7.38])

# 3) Basis definition
basis = ot.ConstantBasisFactory(inputDimension).build()

kriging_result = build_kriging_result(inputSample, outputSample, covarianceModel, basis)
gpr_result = build_gpr_result(inputSample, outputSample, covarianceModel, basis)

# Define a 2D mesh
vertices = [[1.0, 0.0], [2.0, 0.0], [2.0, 1.0], [1.0, 1.0], [1.5, 0.5]]
simplicies = [[0, 1, 4], [1, 2, 4], [2, 3, 4], [3, 0, 4]]
mesh2D = ot.Mesh(vertices, simplicies)

# update the vertices
vertices = ot.Sample(inputSample)
vertices.add(ot.JointDistribution([ot.Uniform(0.0, 10.0)] * 2).getSample(100))

for result in [kriging_result, gpr_result]:
    ot.RandomGenerator.SetSeed(0)
    process = otexp.ConditionedGaussianProcess(result, mesh2D)

    # Get a realization of the process
    realization = process.getRealization()
    print("realization = ", repr(realization))

    # Get a sample & compare it to expectation
    sample = process.getSample(5000)
    mean = sample.computeMean()
    print("Mean over 5000 realizations = ", repr(mean))

    # Check if one can sample the process over a mesh containing conditioning points
    # and 100 new points
    process = otexp.ConditionedGaussianProcess(result, ot.Mesh(vertices))
    realization = process.getRealization()
    num = 0.0
    den = 0.0
    for i in range(len(inputSample)):
        num += (realization.getValueAtIndex(i) - outputSample[i]).norm()
        den += outputSample[i].norm()
    error = num / den
    ott.assert_almost_equal(error, 0.0, 1.0e-6, 1.0e-6)

# 2D use case - #2769
model = ot.SymbolicFunction(["x", "y"], ["cos(x) + sin(y)", "cos(0.5*x) + sin(y)"])
inputSample = ot.Sample([[0.0, 0.0], [0.0, 1.0], [1.0, 0.0], [1.0, 1.0], [0.5, 0.5]])
outputSample = model(inputSample)
covarianceModel = ot.TensorizedCovarianceModel([ot.SquaredExponential([1.0, 1.0])] * 2)

algo = ot.KrigingAlgorithm(inputSample, outputSample, covarianceModel)
algo.run()
result = algo.getResult()
vertices = [[0.3, 0.6], [0.4, 0.8]]
mesh2D = ot.Mesh(vertices)
process = otexp.ConditionedGaussianProcess(result, mesh2D)
sample = process.getSample(3)
ott.assert_almost_equal(sample.getSize(), 3, 0, 0)
ott.assert_almost_equal(sample.getDimension(), 2, 0, 0)

# 2D use case (#2769) with GPR
fitter_algo = otexp.GaussianProcessFitter(inputSample, outputSample, covarianceModel)
fitter_algo.run()
fitter_result = fitter_algo.getResult()
gpr_algo = otexp.GaussianProcessRegression(fitter_result)
gpr_algo.run()
gpr_result = gpr_algo.getResult()
process = otexp.ConditionedGaussianProcess(gpr_result, mesh2D)
ott.assert_almost_equal(sample.getSize(), 3, 0, 0)
ott.assert_almost_equal(sample.getDimension(), 2, 0, 0)

File: lib/test/t_ConditionedGaussianProcess_std.cxx
//                                               -*- C++ -*-
/**
 *  @brief The test file of class ConditionedGaussianProcess for standard methods
 *
 *  Copyright 2005-2020 Airbus-EDF-IMACS-ONERA-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include "openturns/OT.hxx"
#include "openturns/OTtestcode.hxx"

using namespace OT;
using namespace OT::Test;

int main(int, char *[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);

  // Set Numerical precision to 3
  PlatformInfo::SetNumericalPrecision(3);

  // Process with empty constructor
  ConditionedGaussianProcess myProcess;
  fullprint << "process = " << myProcess << std::endl;

  // Learning data
  Point levels(2);
  levels[0] = 8;
  levels[1] = 5;
  // Define the Box
  Box box(levels);
  // Get the input sample
  Sample inputSample( box.generate() );
  // Scale each direction
  inputSample *= 10;

  // Define model
  Description inputDescription(2);
  inputDescription[0] = "x";
  inputDescription[1] = "y";

  Description formula(1);
  formula[0] = "cos(0.5*x) + sin(y)" ;
  const SymbolicFunction model(inputDescription, formula);

  // Build the output sample
  const Sample  outputSample( model(inputSample) );

  // 2) Definition of exponential model
  Point scale(2);
  scale[0] = 1.988;
  scale[1] = 0.924;
  Point amplitude(1, 3.153);
  SquaredExponential covarianceModel(scale, amplitude);

  // 3) Basis definition
  Basis basis(ConstantBasisFactory(2).build());

  // Kriging algorithm
  KrigingAlgorithm algo(inputSample, outputSample, covarianceModel, basis, true);
  algo.run();
  // Get result
  KrigingResult result(algo.getResult());
  // Build a mesh
  // Start with vertices
  Sample vertices(0, 2);
  Point p(2);

  p[0] = 1.0;
  p[1] = 0.0;
  vertices.add(p);

  p[0] = 2.0;
  p[1] = 0.0;
  vertices.add(p);

  p[0] = 2.0;
  p[1] = 1.0;
  vertices.add(p);

  p[0] = 1.0;
  p[1] = 1.0;
  vertices.add(p);

  p[0] = 1.5;
  p[1] = 0.5;
  vertices.add(p);


  IndicesCollection simplices(4, 3);
  simplices(0, 0) = 0;
  simplices(0, 1) = 1;
  simplices(0, 2) = 4;

  simplices(1, 0) = 1;
  simplices(1, 1) = 2;
  simplices(1, 2) = 4;

  simplices(2, 0) = 2;
  simplices(2, 1) = 3;
  simplices(2, 2) = 4;

  simplices(3, 0) = 3;
  simplices(3, 1) = 0;
  simplices(3, 2) = 4;

  Mesh mesh2D(vertices, simplices);


  // Conditioned process
  ConditionedGaussianProcess process(result, mesh2D);
  // Get a realization of the process
  Field realization( process.getRealization() );
  fullprint << "realization = " << realization << std::endl;

  //  Get a sample & compare it to expectation
  ProcessSample sample(process.getSample(5000));
  fullprint << "Mean over 5000 realizations = " << sample.computeMean() << std::endl;

  return ExitCode::Success;
}
